pkg/networkservice/core/eventchannel/monitor_connection_client.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2020 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package eventchannel
+
+import (
+	"context"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/metadata"
+)
+
+type monitorConnectionMonitorConnectionsClient struct {
+	eventCh    <-chan *networkservice.ConnectionEvent
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+}
+
+// NewMonitorConnectionMonitorConnectionsClient - returns a networkservice.MonitorConnection_MonitorConnectionsClient
+//                                                 eventCh - when an event is sent on eventCh, it is returned by the
+//                                                 call to Recv on the networkservice.MonitorConnection_MonitorConnectionsClient
+func NewMonitorConnectionMonitorConnectionsClient(eventCh <-chan *networkservice.ConnectionEvent) networkservice.MonitorConnection_MonitorConnectionsClient {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	return &monitorConnectionMonitorConnectionsClient{
+		eventCh:    eventCh,
+		ctx:        ctx,
+		cancelFunc: cancelFunc,
+	}
+}
+
+func (m *monitorConnectionMonitorConnectionsClient) Recv() (*networkservice.ConnectionEvent, error) {
+	event, ok := <-m.eventCh
+	if !ok {
+		m.cancelFunc()
+		return nil, errors.New("No more events, chan closed by sender")
+	}
+	return event, nil
+}
+
+func (m *monitorConnectionMonitorConnectionsClient) Header() (metadata.MD, error) {
+	return make(metadata.MD), nil
+}
+
+func (m *monitorConnectionMonitorConnectionsClient) Trailer() metadata.MD {
+	return make(metadata.MD)
+}
+
+func (m *monitorConnectionMonitorConnectionsClient) CloseSend() error {
+	return nil
+}
+
+func (m *monitorConnectionMonitorConnectionsClient) Context() context.Context {
+	return m.ctx
+}
+
+func (m *monitorConnectionMonitorConnectionsClient) SendMsg(msg interface{}) error {
```

Comment:
There are several NOP methods here, is this intentional? If not, can we mark them with TODOs?

Reply:
These are intentional NOOP.  The interface was designed for an HTTP/2 intermediary.  These are intentionally designed to be *channel* intermediated.  So these methods just have to *not explode* when called.